File: src/components/robot.py
from math import pi, sin, cos

class RobotState:
    def __init__(self, v = 0, w = 0):
        self.v = v
        self.w = w

class Robot:
    def __init__(self, motion_model):
        self._motion_model = motion_model
        self._sensor = None
        self._robot_state = RobotState()

    def get_pose(self):
        pose = self._motion_model.get_pose()
        return pose

    def set_velocity(self, v):
        self._robot_state.v = v

    def set_angular_velocity(self, w):
        self._robot_state.w = w

    def add_sensor(self, sensor):
        self._sensor = sensor

    def move(self, dt):
        v = self._robot_state.v
        w = self._robot_state.w
        self._motion_model.move(v, w, dt)
        self._update_sensor_pose()

    def _update_sensor_pose(self):
        if self._sensor:
            pose = self.get_pose()
            self._sensor.update_pose(pose.x, pose.y, pose.theta)

    def get_sensor_reading(self):
        sensor = self._sensor
        return sensor.get_sensor_reading()

    def accept(self, visitor):
        visitor.visit(self)

    

File: src/MATLAB/components/Robot.m
classdef Robot
    properties(Access = private)
        motionModel
        sensor
        robotState
    end
    
    methods
        function obj = Robot(motionModel)
            obj.motionModel = motionModel;
            obj.sensor = [];
            obj.robotState = RobotState(0, 0);
        end
        
        function pose = getPose(obj)
            pose = obj.motionModel.getPose();
        end
        
        function v = getVelocity(obj)
            v = obj.robotState.v;
        end
        
        function w = getAngularVelocity(obj)
            w = obj.robotState.w;
        end
        
        function setVelocity(obj, v)
            obj.robotState.v = v;
        end
        
        function setAngularVelocity(obj, w)
            obj.robotState.w = w;
        end
        
        function addSensor(obj, sensor)
            obj.sensor = sensor;
            obj.updateSensorPose();
        end
        
        function move(obj, dt)
            v = obj.robotState.v;
            w = obj.robotState.w;
            obj.motionModel.move(v, w, dt);
            obj.updateSensorPose();
        end
        
        function updateSensorPose(obj)
            if ~isempty(obj.sensor)
                pose = obj.getPose();
                obj.sensor.updatePose(pose.x, pose.y, pose.theta);
            end
        end
        
        function sensorReading = getSensorReading(obj)
            sensorReading = obj.sensor.getSensorReading();
        end
        
        function sensorOffsets = getSensorOffsets(obj)
            sensorOffsets = obj.sensor.getSensorOffsets();
        end
    end
end
